pkg/controller/servicebindingrequest/annotations/resource.go

```diff
@@ -0,0 +1,208 @@
+package annotations
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+
+	"github.com/redhat-developer/service-binding-operator/pkg/controller/servicebindingrequest/nested"
+)
+
+// ResourceHandler handles annotations related to external resources.
+type ResourceHandler struct {
+	// bindingInfo contains the binding details related to the annotation handler.
+	bindingInfo *BindingInfo
+	// client is the client used to retrieve a related secret.
+	client dynamic.Interface
+	// relatedGroupVersionResource is the related resource GVR, used to retrieve the related resource
+	// using the client.
+	relatedGroupVersionResource schema.GroupVersionResource
+	// relatedResourceName is the name of the related resource that is referenced by the handler
+	// annotation key/value pair.
+	relatedResourceName string
+	// resource is the unstructured object to extract data using inputPath.
+	resource unstructured.Unstructured
+	// stringValue is a function used to decode values from the resource being handled; for example,
+	// to decode Base64 keys the decodeBase64String can be used.
+	stringValue func(interface{}) (string, error)
+	// inputPathRoot indicates the root where input paths will be applied to extract a value from the
+	// resource.
+	inputPathRoot *string
+	// restMapper allows clients to map resources to kind, and map kind and version
+	// to interfaces for manipulating those objects.
+	restMapper meta.RESTMapper
+}
+
+// discoverRelatedResourceName returns the resource name referenced by the handler. Can return an
+// error in the case the expected information doesn't exist in the handler's resource object.
+func discoverRelatedResourceName(obj map[string]interface{}, bindingInfo *BindingInfo) (string, error) {
+	resourceNameValue, ok, err := unstructured.NestedFieldCopy(
+		obj,
+		strings.Split(bindingInfo.ResourceReferencePath, ".")...,
+	)
+	if !ok {
+		return "", ResourceNameFieldNotFoundErr
+	}
+	if err != nil {
+		return "", err
+	}
+	name, ok := resourceNameValue.(string)
+	if !ok {
+		return "", InvalidArgumentErr(bindingInfo.ResourceReferencePath)
```

Comment:
Why are we returning `InvalidArgumentErr(bindingInfo.ResourceReferencePath)` if we have error in converting `resourceNameValue` interface to string?

Reply:
An annotation such as `.../status.dbCredentials-password: binding:env:object:configmap`  expresses the intent of a) extract the password field from the ConfigMap named like the value in ` status.dbCredentials`. This value should be only a string, and if it isn't the error is returned.

Given that, the `bindingInfo` parameter could be removed and `resourceReferencePath` introduced to simplify the function and error type could be better.